### PR TITLE
Ability to use X-to-many association sub-properties in Filter component #518

### PR DESCRIPTION
**Pull request by yuriy-khrebtov**
See https://github.com/jmix-framework/jmix/issues/518

### CODE REVIEW

File: jmix-core/core/src/main/java/io/jmix/core/querycondition/PropertyCondition.java [hunk 1]
```diff
@@ -291,5 +313,8 @@ public static class Operation {
         public static final String IN_LIST = "in_list";
         public static final String NOT_IN_LIST = "not_in_list";
         public static final String IN_INTERVAL = "in_interval";
+        public static final String IS_EMPTY = "is_empty";
+        public static final String MEMBER_OF_LIST = "member_of_list";
+        public static final String NOT_MEMBER_OF_LIST = "not_member_of_list";
```
Comment (gorbunkov) [inline jmix-core/core/src/main/java/io/jmix/core/querycondition/PropertyCondition.java:318]: Please create an issue for Studio to support collection attributes for `<propertyFilter/>` component. Currently Studio doesn't suggest collection attributes and mark new operations as not supported for the given entity attribute.

image

File: jmix-ui/ui/src/main/java/io/jmix/ui/component/PropertyFilter.java [hunk 2]
```diff
@@ -184,6 +184,9 @@ enum Operation implements EnumClass<String> {
         IN_LIST(Type.LIST),
         NOT_IN_LIST(Type.LIST),
         DATE_INTERVAL(Type.INTERVAL),
+        IS_EMPTY(Type.UNARY),
```
[the remark this answers is not in the document]
Comment (yuriy-khrebtov) [reply jmix-ui/ui/src/main/java/io/jmix/ui/component/PropertyFilter.java:187]: Looks like "IS_NOT_SET" operation is not handled in the framework, see io.jmix.ui.component.propertyfilter.PropertyFilterSupport#toPropertyConditionOperation. Also, existing operations are mapped to JPQL operations (io.jmix.core.querycondition.PropertyConditionUtils#getJpqlOperation) so it's not possible to assign multiple jpql conditions to single operation.
"IS_COLLECTION_EMPTY" is nice for me.

File: jmix-core/core/src/main/java/io/jmix/core/querycondition/PropertyCondition.java [hunk 3]
```diff
@@ -291,5 +313,8 @@ public static class Operation {
         public static final String IN_LIST = "in_list";
         public static final String NOT_IN_LIST = "not_in_list";
         public static final String IN_INTERVAL = "in_interval";
+        public static final String IS_EMPTY = "is_empty";
+        public static final String MEMBER_OF_LIST = "member_of_list";
```
[the remark this answers is not in the document]
Comment (yuriy-khrebtov) [reply jmix-core/core/src/main/java/io/jmix/core/querycondition/PropertyCondition.java:317]: "LIST" is added to make it similar to "IN_LIST" and "NOT_IN_LIST". But maybe it's not quite correct.

File: jmix-core/core/src/main/java/io/jmix/core/querycondition/PropertyCondition.java (hunk 1 above)
Comment (yuriy-khrebtov) [reply jmix-core/core/src/main/java/io/jmix/core/querycondition/PropertyCondition.java]: https://youtrack.haulmont.com/issue/JST-4537/Support-x-to-many-property-conditions-of-filter-component

**Review by gorbunkov — APPROVED** [review]

**Pull request merged**